neuro_san/internals/messages/intra_agent_message_utils.py

```diff
@@ -0,0 +1,98 @@
+
+# Copyright (C) 2023-2025 Cognizant Digital Business, Evolutionary AI.
+# All Rights Reserved.
+# Issued under the Academic Public License.
+#
+# You can be released from the terms, and requirements of the Academic Public
+# License by purchasing a commercial license.
+# Purchase of a commercial license is mandatory for any use of the
+# neuro-san SDK Software in commercial settings.
+#
+# END COPYRIGHT
+
+from typing import Any
+from typing import List
+from typing import Union
+
+import json
+
+from langchain_core.messages.base import BaseMessage
+
+from neuro_san.internals.messages.chat_message_type import ChatMessageType
+
+
+class IntraAgentMessageUtils:
+    """
+    Utility class to parse messages with role information.
+    This ancient style of message is used between agents and tool calls
+    and is agnostic to OpenAI vs langchain.
+    """
+
+    @staticmethod
+    def generate_response(the_messages: List[Any]) -> str:
+        """
+        :param the_messages: A list of messages. Could be OpenAI or langchain messages.
+        :return: a JSON-ification of the list of messages.
+        """
+        response_list = []
+        for index, one_message in enumerate(the_messages):
+            # Duplicate the role message before every tool response role message
+            role: str = IntraAgentMessageUtils._get_role(one_message)
+            if role == "tool" and index > 0:
+                previous_message = the_messages[index - 1]
+                previous_role: str = IntraAgentMessageUtils._get_role(previous_message)
+                if previous_role == "assistant":
+                    new_message = {
+                        "role": previous_role,
+                        "content": IntraAgentMessageUtils.get_content(previous_message)
+                    }
+                    response_list.append(new_message)
+
+            message_dict = {
+                "role": role,
+                "content": IntraAgentMessageUtils.get_content(one_message)
+            }
+            response_list.append(message_dict)
+
+        return json.dumps(response_list)
+
+    @staticmethod
+    def _get_role(message: Any) -> str:
```

Comment:
This comes from message_utils.py and is made a static method on a class.
This is also made private since no one else uses it